Add a function to get the size of a collection/table in bytes.

faisalill opened the pull request:
### What's Changed:
Added `getCollectionSize` method to MariaDB, MySQL, SQLite and MongoDB adapters.
The function for Postgres is working but it is giving an error during test so I just commented the code out.
 ### Tests:
 I have checked manually using the `getCollectionSize` method it worked.
 I did add a test to check whether it returns a string.

faisalill commented:
@abnegate

abnegate commented on README.md:
````diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
+$database = new Database(new MariaDB($pdo), $cache);
+```
+
+**MySql:**
````
Let's stylize with capitals

```suggestion
**MySQL:**
```

abnegate commented on README.md:
```diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
+use Utopia\Database\Adapter\MySQL;
 
-$dbHost = 'mariadb';
+$dbHost = 'MySqlHost ';
```
Let' just use `mysql` as the host here

abnegate commented on README.md:
```diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
+    PDO::ATTR_STRINGIFY_FETCHES => true,
+];
+
+$pdo = new PDO("mysql:host={$dbHost};port={$dbPort};charset=utf8mb4", $dbUser, $dbPass, $pdoConfig );
```
```suggestion
$pdo = new PDO("mysql:host={$dbHost};port={$dbPort};charset=utf8mb4", $dbUser, $dbPass, $pdoConfig);
```

abnegate commented on README.md:
```diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
+    PDO::ATTR_STRINGIFY_FETCHES => true,
+];
+
+$pdo = new PDO("mysql:host={$dbHost};port={$dbPort};charset=utf8mb4", $dbUser, $dbPass, $pdoConfig );
```
```suggestion
$pdo = new PDO("mysql:host={$dbHost};port={$dbPort};charset=utf8mb4", $dbUser, $dbPass, $pdoConfig);
```

abnegate commented on README.md:
```diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
-]);
+];
+
+$pdo = new PDO("pgsql:host={$dbHost};port={$dbPort};", $dbUser, $dbPass, $pdoConfig );
```
```suggestion
$pdo = new PDO("pgsql:host={$dbHost};port={$dbPort};charset=utf8mb4", $dbUser, $dbPass, $pdoConfig);
```

abnegate commented on README.md:
```diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
+use Utopia\Cache\Adapter\None as NoCache;
+use Utopia\Database\Adapter\Postgres;
+
+$dbHost = 'PostgresHost ';
```
Let's just use `postgres` for the host here

abnegate commented on README.md:
```diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
+use Utopia\Database\Adapter\Postgres;
+
+$dbHost = 'PostgresHost ';
+$dbPort = '3306';
```
Let's use `5432` for the port as it's the default for Postgres

abnegate commented on README.md:
```diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
+
+**MongoDB:**
+
+> This requires [utopia-php/mongo](https://packagist.org/packages/utopia-php/mongo)
```
Let's remove this, it's already a dependency of this library so end users won't need to depend on it too

abnegate commented on README.md:
```diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
+use Utopia\Database\Adapter\Mongo;
+use Utopia\Mongo\Client; // from utopia-php/mongo
+
+$dbHost = 'MySqlHost ';
```
Let's use `mongo` as the host here

abnegate commented on README.md:
```diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
+use Utopia\Mongo\Client; // from utopia-php/mongo
+
+$dbHost = 'MySqlHost ';
+$dbPort = 3306; // this should be a integer
```
Let's use `27017` as the port here as it's the default for Mongo, and remove the comment

abnegate commented on README.md:
```diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
+$dbUser = 'root';
+$dbPass = 'password';
+$dbName = 'dbName';
+$useCoroutine = true; // set to false if you don't want to use Swoole\Coroutine(set true for concurrent requests and better performance)
```
Let's remove this declaration and hard-code `useCoroutine: true` in the client constructor

abnegate commented on README.md:
```diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
+$dbPass = 'password';
+$dbName = 'dbName';
+$useCoroutine = true; // set to false if you don't want to use Swoole\Coroutine(set true for concurrent requests and better performance)
+
```
Let's remove this blank line

abnegate commented on README.md:
````diff
@@ -43,40 +43,145 @@ Attribute filters are functions that manipulate attributes before saving them to
+```php
+$nameOfTheDatabaseOrSchema = 'mydb';
+$database->setNamespace($nameOfTheDatabaseOrSchema);
+$database->create(); // Creates a new database for MySql, MariaDB, SQLite. For Postgres it creates a schema. named 'mydb'
````
Let's add the comment above the line of code instead of after

abnegate reviewed (commented):
Great start :grin:

faisalill commented:
@abnegate ,

1.Fixed the stuff u mentioned. 

2.As for postgres and sqlite I checked and changed those queries they work  but in tests they give error.
![getSIzeOfCollectionErrors](https://github.com/utopia-php/database/assets/118456685/9efa903d-9286-410b-8fcc-f56b3641b0b9)

3. Regarding the Documentation, I guess only Document and Relationship methods are pending.

4. Can I get permission to add try catch blocks which throws Database Exception as there are a lot of methods in the adapters that are missing it.

faisalill commented on src/Database/Adapter.php:
```diff
@@ -186,6 +186,16 @@ abstract public function delete(string $name): bool;
+     * @throws Exception
+     */
+     
+     abstract public function getCollectionSize(string $name): string; 
```
Done

faisalill commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,43 @@ public function createCollection(string $name, array $attributes = [], array $in
+     
+         $query = $this->getPDO()->prepare("
+             SELECT 
+              data_length + index_length + data_free
```
Changed it to 
data_length + index_length

faisalill commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,43 @@ public function createCollection(string $name, array $attributes = [], array $in
+         ");
+         $query->bindParam(':database', $database);
+         $query->bindParam(':name', $name);
+         $query->execute();
```
Done

faisalill commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,43 @@ public function createCollection(string $name, array $attributes = [], array $in
+            $query->execute();
+            $size = $query->fetchColumn();
+        } catch (PDOException $e) {
+            throw new Exception($e->getMessage());
```
Done. Did it for other Adapters too.

faisalill commented on src/Database/Adapter/Postgres.php:
```diff
@@ -148,6 +148,37 @@ public function createCollection(string $name, array $attributes = [], array $in
+    //     } catch (PDOException $e) {
+    //         throw new Exception( $e->getMessage());
+    //     }
+    //      return (int) $size;
```
It works but the problem is in tests.

faisalill commented on src/Database/Adapter/SQLite.php:
```diff
@@ -158,6 +158,34 @@ public function createCollection(string $name, array $attributes = [], array $in
+    $query = $this->getPDO()->prepare("
+        SELECT sum(LENGTH(name) + LENGTH(sql)) AS total_size
+        FROM sqlite_master
+        WHERE type = 'table' AND name = '{$name}'
```
Fixed it

faisalill commented on src/Database/Adapter/SQLite.php:
```diff
@@ -158,6 +158,34 @@ public function createCollection(string $name, array $attributes = [], array $in
+        FROM sqlite_master
+        WHERE type = 'table' AND name = '{$name}'
+    ");
+    $query->execute();
```
Done

faisalill commented on src/Database/Adapter/Mongo.php:
```diff
@@ -253,6 +253,32 @@ public function listCollections(): array
+             'scale' => 1 
+         ];
+     
+         $result = $this->getClient()->query($command, $database);
```
Done

faisalill commented on tests/Database/Base.php:
```diff
@@ -113,6 +113,8 @@ public function testCreateListExistsDeleteCollection(): void
         $this->assertEquals(true, static::getDatabase()->deleteCollection('actors2')); // Delete collection when finished
         $this->assertCount(2, static::getDatabase()->listCollections());
 
+        $this->assertIsString(static::getDatabase()->getCollectionSize("actors"));
```
It is returning 0 i don't know why. Testing it before and after adding data

faisalill commented on src/Database/Adapter.php:
```diff
@@ -186,6 +186,16 @@ abstract public function delete(string $name): bool;
+     * @return int
+     * @throws Exception
+     */
+     
```
Done

abnegate commented:
@faisalill 

> 1. Fixed the stuff u mentioned.

Awesome, nice work!

> 2.  For the testing part I don't know why but it keeps returning me 0 (checked both before and after adding data) . That's the reason I used assertIsInt. Postgres is giving me the collection size in bytes but during test it is giving me an error SQLite too but it's related to the function returning null/0 as I said.

Please try to dig into this further, we'll need to get all adapters working. How are you running the tests? The Postgres issue is strange for the test suite. For the SQLite issue, try some extra debugging, adding `var_dump` or other logging to see what the function is doing at each step, that should help figure out why you're getting null. 

> 3. Regarding the Documentation, I guess only Document and Relationship methods are pending.

Awesome, let's add those methods too

> 4. Can I get permission to add try catch blocks which throws Database Exception as there are a lot of methods in the adapters that are missing it.

Let's address that seperately, it's outside the scope of this PR. Make sure you've merge in the upstream main branch too, a lot of the `Exception`'s have been updated to `DatabaseException`. DM me if you need help with it

faisalill commented:
@abnegate, 
1. Completed the documentation.
2. Got the tests working.
3. Changed the tests. (I check whether the size before and after data is greater)
Thank you

abnegate commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,40 @@ public function createCollection(string $name, array $attributes = [], array $in
+         $query->bindParam(':database', $database);
+         $query->bindParam(':name', $collectionName);
+         try {
+            $query->execute();
+            $size = $query->fetchColumn();
+        } catch (PDOException $e) {
+            throw new DatabaseException('Failed to get collection size: ' . $e->getMessage());
+        }
```
Let's make sure all the indentation is consistent, you can use the linter to do that. From the project root, run either:

```bash
composer format
```
or if you don't have composer installed locally:
```bash
docker run -itv $PWD:/app composer format
```

abnegate commented on src/Database/Adapter/Mongo.php:
```diff
@@ -253,6 +253,28 @@ public function listCollections(): array
+            $result = $this->getClient()->query($command);
+        }
+        catch(Exception $e){
+            throw new DatabaseException("Failed to get collection size" . $e->getMessage());
```
Let's use single quotes if we're not using string interpolation as it is faster, and also make sure our error messages are consistent across adapters

```suggestion
            throw new DatabaseException('Failed to get collection size: ' . $e->getMessage());
```

abnegate commented on tests/Database/Base.php:
```diff
@@ -118,6 +118,70 @@ public function testCreateListExistsDeleteCollection(): void
+
+            $valueBeforeAddingData = static::getDatabase()->getSizeOfCollection('size_testing');
+
+            static::getDatabase()->createAttribute('size_testing', 'string1', Database::VAR_STRING, 128, true);
+            static::getDatabase()->createAttribute('size_testing', 'string2', Database::VAR_STRING, 2147483646 + 1, true);
+            static::getDatabase()->createAttribute('size_testing', 'string3', Database::VAR_STRING, 2147483646 + 1, true);
+            static::getDatabase()->createAttribute('size_testing', 'string4', Database::VAR_STRING, 2147483646 + 1, true);
+            static::getDatabase()->createAttribute('size_testing', 'integer', Database::VAR_INTEGER, 0, true);
+            static::getDatabase()->createAttribute('size_testing', 'bigint', Database::VAR_INTEGER, 8, true);
+            static::getDatabase()->createAttribute('size_testing', 'float', Database::VAR_FLOAT, 0, true);
+            static::getDatabase()->createAttribute('size_testing', 'boolean', Database::VAR_BOOLEAN, 0, true);
+            static::getDatabase()->createAttribute('size_testing', 'string1new', Database::VAR_STRING, 128, false);
+            static::getDatabase()->createAttribute('size_testing', 'string2new', Database::VAR_STRING, 2147483646 + 1, false);
+            static::getDatabase()->createAttribute('size_testing', 'string3new', Database::VAR_STRING, 2147483646 + 1, false);
+            static::getDatabase()->createAttribute('size_testing', 'string4new', Database::VAR_STRING, 2147483646 + 1, false);
+            static::getDatabase()->createAttribute('size_testing', 'integernew', Database::VAR_INTEGER, 0, false);
+            static::getDatabase()->createAttribute('size_testing', 'bigintnew', Database::VAR_INTEGER, 8, false);
+            static::getDatabase()->createAttribute('size_testing', 'floatnew', Database::VAR_FLOAT, 0, false);
+            static::getDatabase()->createAttribute('size_testing', 'booleannew', Database::VAR_BOOLEAN, 0, false);
+            static::getDatabase()->createIndex('size_testing', 'string1_index', Database::INDEX_KEY, ['string1']);
+            static::getDatabase()->createIndex('size_testing', 'string2_index', Database::INDEX_KEY, ['string2'], [255]);
+            static::getDatabase()->createIndex('size_testing', 'multi_index', Database::INDEX_KEY, ['string1', 'string2', 'string3'], [128, 128, 128]);
```
Let's try create the least amount of attributes we need to test the size change

abnegate commented on tests/Database/Base.php:
```diff
@@ -118,6 +118,70 @@ public function testCreateListExistsDeleteCollection(): void
+            for ($i = 0; $i < 100; $i++) {
+                static::getDatabase()->createDocument('size_testing', new Document([
+                    '$permissions' => [
+                        Permission::read(Role::any()),
+                        Permission::read(Role::user(ID::custom('1'))),
+                        Permission::read(Role::user(ID::custom('2'))),
+                        Permission::create(Role::any()),
+                        Permission::create(Role::user(ID::custom('1x'))),
+                        Permission::create(Role::user(ID::custom('2x'))),
+                        Permission::update(Role::any()),
+                        Permission::update(Role::user(ID::custom('1x'))),
+                        Permission::update(Role::user(ID::custom('2x'))),
+                        Permission::delete(Role::any()),
+                        Permission::delete(Role::user(ID::custom('1x'))),
+                        Permission::delete(Role::user(ID::custom('2x'))),
```
Let's use the least amount of permissions required for the size check so that they're not distracting from the purpose of the test case

abnegate commented on tests/Database/Base.php:
```diff
@@ -118,6 +118,70 @@ public function testCreateListExistsDeleteCollection(): void
+
+              $valueAfterAddingData = static::getDatabase()->getSizeOfCollection('size_testing');
+     
+            if(static::getAdapterName() === 'mysql'){
+                $this->assertGreaterThan(65535,$valueAfterAddingData);
+            }
+            else{
+                $this->assertGreaterThan($valueBeforeAddingData,$valueAfterAddingData);
+            }
```
Can we make this the same for all adapters instead of splitting for mysql?

faisalill replied:
I tried to add different attributes and almost 10000 documents but still mysql keeps returning me 65536 as the size. I checked the size using sql terminal and it still returns me 65536. Adding so many documents is just increasing the time of tests. Rest all the adapters are working fine.

faisalill commented:
@abnegate,
1. Formatted  using `composer format` for indentation.
2. Changed the amount of attributes created for testing.
3. Regarding test for mysql : I tried to add different attributes and almost 10000 documents but still mysql keeps returning me 65536 as the size. I checked the size using sql terminal and it still returns me 65536. Adding so many documents is just increasing the time of tests. Rest all the adapters are working fine their size changes but mysql.

abnegate commented on src/Database/Adapter/SQLite.php:
```diff
@@ -158,13 +158,37 @@ public function createCollection(string $name, array $attributes = [], array $in
+         $name = $this->filter($collection);
+         $namespace = $this->getNamespace();
+         $query = $this->getPDO()->prepare("
+    SELECT SUM(\"pgsize\") FROM \"dbstat\" WHERE name='{$namespace}_{$name}';
+    ");
```
Let's fix the indentation here

abnegate commented on src/Database/Adapter/Mongo.php:
```diff
@@ -253,13 +253,34 @@ public function listCollections(): array
+         } catch(Exception $e) {
+             throw new DatabaseException('Failed to get collection size: ' . $e->getMessage());
+         }
+          return $result->size;
```
We still have some indentation issues here

abnegate commented on src/Database/Adapter/Postgres.php:
```diff
@@ -148,12 +148,35 @@ public function createCollection(string $name, array $attributes = [], array $in
+         } catch (PDOException $e) {
+             throw new DatabaseException('Failed to get collection size: ' . $e->getMessage());
+         }
+          return  $size;
```
We still have some indentation issues here

abnegate commented on src/Database/Adapter/SQLite.php:
```diff
@@ -158,13 +158,37 @@ public function createCollection(string $name, array $attributes = [], array $in
+         } catch (PDOException $e) {
+             throw new DatabaseException('Failed to get collection size: ' . $e->getMessage());
+         }
+         return $size;
```
Let's add a line break before the `return`'s

abnegate commented on tests/Database/Base.php:
```diff
@@ -119,6 +119,49 @@ public function testCreateListExistsDeleteCollection(): void
+
+        $valueAfterAddingData = static::getDatabase()->getSizeOfCollection('size_testing');
+
+        if (static::getAdapterName() === 'mysql') {
+            $this->assertGreaterThan(65535, $valueAfterAddingData);
+        } else {
+            $this->assertGreaterThan($valueBeforeAddingData, $valueAfterAddingData);
+        }
```
Is there any amount of documents/attributes that changes this for MySQL? We need to find a way to make the assertion the same there so we don't switch on the adapter name

abnegate commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,39 @@ public function createCollection(string $name, array $attributes = [], array $in
+             throw new DatabaseException('Failed to get collection size: ' . $e->getMessage());
+         }
+
+          return $size;
```
Let's fix the indentation here

faisalill replied:
Yep fixed it

faisalill commented:
@abnegate , 
Fixed the indentation as you mentioned.

faisalill commented:
@abnegate , 
The tests are working fine now.
Can you re-review please

fogelito commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,39 @@ public function createCollection(string $name, array $attributes = [], array $in
+    public function getSizeOfCollection(string $collection): int
+    {
+        $name = $this->filter($collection);
+        $collectionName = "{$this->getNamespace()}_{$name}";
```
For consistency let's use the getSQLTable()

faisalill replied:
Is the following code okay??

```php
$database = str_replace('`', '', explode('.', $this->getSQLTable($name))[0]);
$collectionName = str_replace('`', '', explode('.', $this->getSQLTable($name))[1]);
 ```

fogelito replied:
The $database from $this->getNamespace() is ok.
But because of the str_replace on the Backticks and since the value is bind, let's resolve the conversation 👍

fogelito commented on tests/Database/Base.php:
```diff
@@ -120,6 +120,53 @@ public function testCreateListExistsDeleteCollection(): void
+        static::getDatabase()->createAttribute('size_test2', 'string1', Database::VAR_STRING, 128, true);
+        static::getDatabase()->createAttribute('size_test2', 'string2', Database::VAR_STRING, 254 + 1, true);
+        static::getDatabase()->createAttribute('size_test2', 'string3', Database::VAR_STRING, 254 + 1, true);
+        static::getDatabase()->createAttribute('size_test2', 'string4', Database::VAR_STRING, 254 + 1, true);
```
Hmmm.. for MySQL it looks like it does not update on the fly. After running optimize table it did, obviously, we will not run such a command.
@abnegate we should remember the value is ~size not exact.
Not sure how the rest adapter work. could be the same.

abnegate replied:
Is there a way we can make it always up-to-date for MySQL?

faisalill replied:
@fogelito @abnegate 
I added the following code in MariaDB class which runs before getting the size. After using it the size changes even if I create one index. 
 ```php
 $analyzeTable = $this->getPDO()->prepare("ANALYZE TABLE {$database}.{$collectionName}");
$analyzeTable->execute();
```

fogelito replied:
Show variables like 'information_schema_stats_expiry' is 86400 (1 day).
The data is cached for a day.
Looks like @faisalill is right 👍 as said here:
https://dev.mysql.com/doc/refman/8.0/en/information-schema-tables-table.html
Looks like ANALYZE TABLE is relatively a fast operation, tried now on 5.1 million table rows.
The question is how often does it run?

faisalill replied:
@fogelito @abnegate Since it analyzes the table everytime the `getSizeOfCollection` function is called which won't be so feasible in production environment. How about I create something like `analyzeDatabase` or a `analyzeCollection` function in the database adapter classes. So the user can analyze the database or collection not just for getting accurate size but to update the stats too. 

Thanks for your time.

abnegate replied:
> Looks like ANALYZE TABLE is relatively a fast operation, tried now on 5.1 million table rows.

@fogelito How long did it take? How small can we make `information_schema_stats_expiry`?

@faisalill Yes please add a separate `analyzeCollection` function and let's call this from the tests before getting the size 👍

fogelito replied:
@abnegate The Analyze took 0.135 sec.
"information_schema_stats_expiry" can be down to 0 , not sure if we want this, since I do not know what will be the effects. On cloud, are we able to change variables in global scope?

faisalill replied:
@fogelito @abnegate 
- Added `analyzeCollection` function to different adapteres (no operation function for MongoDB).
- Working with the same results (adding just one index changes the size except for sqlite which still requires 24 docs).
- All Tests passed

Thank you for your time

fogelito commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -154,6 +154,9 @@ public function getSizeOfCollection(string $collection): int
         $collectionName = "{$this->getNamespace()}_{$name}";
         $database = $this->getDefaultDatabase();
 
+        $analyzeTable = $this->getPDO()->prepare("ANALYZE TABLE {$database}.{$collectionName}");
```
Maybe now we need to use the getSQLTable :)

faisalill replied:
@fogelito used `getSQLTable` this time works fine.

Thank you for your time

abnegate commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,62 @@ public function createCollection(string $name, array $attributes = [], array $in
+        $database = $this->getDefaultDatabase();
+
+        $query = $this->getPDO()->prepare("
+             SELECT SUM(data_length + index_length) AS total_size 
+             FROM information_schema.TABLES 
+             WHERE table_schema = :database
+             AND table_name = :name
```
After some internal investigation, we've found that this doesn't give you an accurate representation of the size of the collection _on disk._ Let's update this query to the following:

```sql
SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
FROM INFORMATION_SCHEMA.INNODB_TABLESPACES
WHERE NAME = CONCAT(:database, '/', :name)
```

fogelito replied:
If we are going to use INNODB_TABLESPACES...
Need to double check, but I think using reading from INNODB_TABLESPACES is precise and we do not need the analyzed table?

Lets bind the Param for NAME.
We always want to send to the query a value, calculated by PHP if possible, It can make the query not use the index.

faisalill replied:
@fogelito @abnegate 
Using the above query mentioned by @abnegate does not require analyze table. 

Note: I used `FROM INFORMATION_SCHEMA.INNODB_SYS_TABLESPACES` for mariadb (https://mariadb.com/kb/en/information-schema-innodb_sys_tablespaces-table/) and `FROM INFORMATION_SCHEMA.INNODB_TABLESPACES` for MySQL.

### Size Change:
MySQL, MariaDB, Postgres: 1 index
MongoDB: 1 Document
SQLite: 26 Documents

abnegate replied:
@faisalill Awesome! As @fogelito mentioned, let's concatenate the entire `NAME` as a string in PHP and bind that instead of using `CONCAT` in the query.

faisalill replied:
@abnegate 
Added the changes as you said.

faisalill replied:
@abnegate 
Added the changes as you said.

abnegate commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,62 @@ public function createCollection(string $name, array $attributes = [], array $in
+     * @return bool
+     * @throws DatabaseException
+     */
+    public function analyzeCollection(string $collection): bool
```
After further thought, let's revert and remove this function and run the analyze directly inside the `getSizeOfCollection` function adapters that require it, since this function only makes sense for adapters that support an analyze query

abnegate commented on tests/Database/Base.php:
```diff
@@ -248,6 +248,39 @@ public function testCreateListExistsDeleteCollection(): void
+            static::getDatabase()->createDocument('size_test2', new Document([
+                '$permissions' => [
+                    Permission::read(Role::any()),
+                    Permission::create(Role::any()),
```
Let's remove the `create` permission here as it only makes sense for collections not documents

abnegate commented on tests/Database/Base.php:
```diff
@@ -248,6 +248,39 @@ public function testCreateListExistsDeleteCollection(): void
+        static::getDatabase()->createCollection('size_test1');
+        static::getDatabase()->createCollection('size_test2');
+ 
+        $size_test_1 = static::getDatabase()->getSizeOfCollection('size_test1');
```
Let's use camelCase for the variable names here

abnegate commented on tests/Database/Base.php:
```diff
@@ -248,6 +248,39 @@ public function testCreateListExistsDeleteCollection(): void
+        $size_test_1 = static::getDatabase()->getSizeOfCollection('size_test1');
+         
+        static::getDatabase()->createAttribute('size_test2', 'string1', Database::VAR_STRING, 128, true);
+        static::getDatabase()->createAttribute('size_test2', 'string2', Database::VAR_STRING, 254 + 1, true);
```
Let's use 255 instead of 254 + 1

abnegate commented on tests/Database/Base.php:
```diff
@@ -248,6 +248,39 @@ public function testCreateListExistsDeleteCollection(): void
 
+    public function testSizeCollection(): void
+    {
+        static::getDatabase()->createCollection('size_test1');
+        static::getDatabase()->createCollection('size_test2');
```
Let's assert these are equal to start with

abnegate commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,38 @@ public function createCollection(string $name, array $attributes = [], array $in
+        try {
+             $query->execute();
+             $size = $query->fetchColumn();
+        }
+        catch (PDOException $e) {
```
Catch should be on the same line as the closing brace

```suggestion
        } catch (PDOException $e) {
```

faisalill replied:
done

abnegate commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,38 @@ public function createCollection(string $name, array $attributes = [], array $in
+        $query = $this->getPDO()->prepare("
+             SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
+             FROM INFORMATION_SCHEMA.INNODB_SYS_TABLESPACES
+             WHERE NAME = CONCAT(:name)
```
We don't need the concat anymore since we build the entire name in PHP:

```suggestion
             WHERE NAME = :name
```

faisalill replied:
Done

abnegate commented on src/Database/Adapter/MySQL.php:
```diff
@@ -78,4 +78,36 @@ protected function processException(PDOException $e): void
+        $query = $this->getPDO()->prepare("
+             SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
+             FROM INFORMATION_SCHEMA.INNODB_TABLESPACES
+             WHERE NAME = CONCAT(:database, '/', :name)
+         ");
+
+        $query->bindParam(':database', $database);
+        $query->bindParam(':name', $collectionName);
```
Let's build the name in PHP as we do for MariaDB here

abnegate commented on src/Database/Adapter/MySQL.php:
```diff
@@ -78,4 +78,36 @@ protected function processException(PDOException $e): void
+             $query->execute();
+             $size = $query->fetchColumn();
+        }
+        catch (PDOException $e) {
```
Catch should be on the same line as the closing brace

abnegate commented on src/Database/Adapter/Postgres.php:
```diff
@@ -151,6 +151,32 @@ public function createCollection(string $name, array $attributes = [], array $in
+             $query->execute();
+             $size = $query->fetchColumn();
+        }
+        catch (PDOException $e) {
```
Catch should be on the same line as the closing brace

abnegate commented on src/Database/Adapter/SQLite.php:
```diff
@@ -158,6 +158,33 @@ public function createCollection(string $name, array $attributes = [], array $in
+             $query->execute();
+             $size = $query->fetchColumn();
+        }
+        catch (PDOException $e) {
```
Catch should be on the same line as the closing brace

faisalill replied:
Done

abnegate commented on src/Database/Adapter/SQLite.php:
```diff
@@ -158,6 +158,33 @@ public function createCollection(string $name, array $attributes = [], array $in
+        $namespace = $this->getNamespace();
+        
+        $query = $this->getPDO()->prepare("
+             SELECT SUM(\"pgsize\") FROM \"dbstat\" WHERE name='{$namespace}_{$name}';
```
Let's build the `name` in PHP and bind it here

faisalill replied:
Yep and did the same for mongo db too.

faisalill replied:
Done

faisalill replied:
Done

faisalill commented on src/Database/Adapter/Mongo.php:
```diff
@@ -253,6 +253,30 @@ public function listCollections(): array
+        try {
+             $result = $this->getClient()->query($command);
+        } 
+        catch(Exception $e) {
```
Done

abnegate commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,46 @@ public function createCollection(string $name, array $attributes = [], array $in
+         ");
+
+        $permissionsSize = $this->getPDO()->prepare("
+         SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
+         FROM INFORMATION_SCHEMA.INNODB_SYS_TABLESPACES
+         WHERE NAME = :permissions
```
Let's fix the indentation here

abnegate replied:
```suggestion
            SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
            FROM INFORMATION_SCHEMA.INNODB_SYS_TABLESPACES
            WHERE NAME = :permissions
```

abnegate replied:
```suggestion
            SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
            FROM INFORMATION_SCHEMA.INNODB_SYS_TABLESPACES
            WHERE NAME = :permissions
```

abnegate commented on src/Database/Adapter/Mongo.php:
```diff
@@ -253,6 +253,35 @@ public function listCollections(): array
+
+        try {
+            $result = $this->getClient()->query($command);
+            if (is_object($result) && property_exists($result, 'size')) {
+                return $result->size;
+            } else {
+                throw new DatabaseException('Failed to get collection size');
+            }
```
Let's remove this check and return `$result->size` directly. The existing catch block will catch any exception thrown (otherwise we will throw/catch/throw and the message will be `Failed to get collection size: Failed to get collection size`)

faisalill replied:
![image](https://github.com/utopia-php/database/assets/118456685/d4088335-599a-4604-b877-a3b7a83de009)

faisalill replied:
Does not pass phpstan check.
If an error is thrown it passes phpstan check.

abnegate replied:
I see, in that case let's instead just change the message to `No size found` so the wrapping exception's message sounds better

abnegate commented on src/Database/Adapter/MySQL.php:
```diff
@@ -78,4 +78,45 @@ protected function processException(PDOException $e): void
+             SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
+             FROM INFORMATION_SCHEMA.INNODB_TABLESPACES
+             WHERE NAME = :permissions
+            ");
```
Let's fix the indentation here

abnegate replied:
```suggestion
        ");
```

abnegate commented on src/Database/Adapter/MySQL.php:
```diff
@@ -78,4 +78,45 @@ protected function processException(PDOException $e): void
+             FROM INFORMATION_SCHEMA.INNODB_TABLESPACES
+             WHERE NAME = :permissions
+            ");
+
```
```suggestion
```

abnegate commented on src/Database/Adapter/MySQL.php:
```diff
@@ -78,4 +78,45 @@ protected function processException(PDOException $e): void
+             SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
+             FROM INFORMATION_SCHEMA.INNODB_TABLESPACES
+             WHERE NAME = :name
+         ");
```
```suggestion
       ");
```

abnegate commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,46 @@ public function createCollection(string $name, array $attributes = [], array $in
+         SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
+         FROM INFORMATION_SCHEMA.INNODB_SYS_TABLESPACES
+         WHERE NAME = :permissions
+         ");
```
```suggestion
        ");
```

abnegate commented on src/Database/Adapter/Mongo.php:
```diff
@@ -253,6 +253,35 @@ public function listCollections(): array
+            if (is_object($result) && property_exists($result, 'size')) {
+                return $result->size;
+            } else {
+                throw new DatabaseException('Failed to get collection size');
```
```suggestion
                throw new DatabaseException('No size found');
```

fogelito commented:
I think we missed the full-text index calculations table size? since there are additional tables.
cc @abnegate

fogelito commented:
looks like the first FT will create 6 table:
6 table:
 appwrite/fts_00000000000004a9_00000000000001ca_index_1
appwrite/fts_00000000000004a9_00000000000001ca_index_2
appwrite/fts_00000000000004a9_00000000000001ca_index_3
appwrite/fts_00000000000004a9_00000000000001ca_index_4
appwrite/fts_00000000000004a9_00000000000001ca_index_5
appwrite/fts_00000000000004a9_00000000000001ca_index_6

with 5 additional table:

appwrite/fts_00000000000004a9_deleted_cache
appwrite/fts_00000000000004a9_deleted
appwrite/fts_00000000000004a9_config
appwrite/fts_00000000000004a9_being_deleted_cache
appwrite/fts_00000000000004a9_being_deleted

The second fulltext index for the same table will create only 6 tables

abnegate commented:
@fogelito Any ideas how we can count those too?

abnegate commented:
@faisalill Please add another test that:

1. Creates a collection
2. Creates some documents
3. Gets the size
4. Adds a fulltext index to the collection
5. Checks that the size is larger than before

faisalill commented:
@abnegate 
Added test for full text indexing it worked for all adapters except for mongoDB.
MongoDB is returning the same size before and after creating an index.

faisalill commented:
@abnegate @fogelito 
Size increases when full text index is added.
Added tests for it too.

abnegate commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,46 @@ public function createCollection(string $name, array $attributes = [], array $in
+        $permissions = $database . '/' . $collection . '_perms';
+
+        $collectionSize = $this->getPDO()->prepare("
+                SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
+                FROM INFORMATION_SCHEMA.INNODB_SYS_TABLESPACES
+                WHERE NAME = :name
```
```suggestion
            SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
            FROM INFORMATION_SCHEMA.INNODB_SYS_TABLESPACES
            WHERE NAME = :name
```

abnegate commented on src/Database/Adapter/MariaDB.php:
```diff
@@ -142,6 +142,46 @@ public function createCollection(string $name, array $attributes = [], array $in
+         ");
+
+        $permissionsSize = $this->getPDO()->prepare("
+                SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
+                FROM INFORMATION_SCHEMA.INNODB_SYS_TABLESPACES
+                WHERE NAME = :permissions
```
```suggestion
            SELECT SUM(FS_BLOCK_SIZE + ALLOCATED_SIZE)  
            FROM INFORMATION_SCHEMA.INNODB_SYS_TABLESPACES
            WHERE NAME = :permissions
```

abnegate commented on src/Database/Adapter/Mongo.php:
```diff
@@ -253,6 +253,35 @@ public function listCollections(): array
+            if (is_object($result)) {
+                return $result->totalSize;
+            } else {
+                throw new DatabaseException("No size found");
```
```suggestion
                throw new DatabaseException('No size found');
```

abnegate commented on tests/Database/Base.php:
```diff
@@ -248,6 +248,76 @@ public function testCreateListExistsDeleteCollection(): void
+
+    public function testSizeFullText(): void
+    {
+        // SQLite does not support fulltext indexes
+        if (static::getDatabase()->getAdapter()->getSupportForFulltextIndex()) {
+            static::getDatabase()->createCollection('fullTextSizeTest');
+
+            $size1 = static::getDatabase()->getSizeOfCollection('fullTextSizeTest');
+
+            static::getDatabase()->createAttribute('fullTextSizeTest', 'string1', Database::VAR_STRING, 128, true);
+            static::getDatabase()->createAttribute('fullTextSizeTest', 'string2', Database::VAR_STRING, 254, true);
+            static::getDatabase()->createAttribute('fullTextSizeTest', 'string3', Database::VAR_STRING, 254, true);
+            static::getDatabase()->createIndex('fullTextSizeTest', 'index', Database::INDEX_KEY, ['string1', 'string2', 'string3'], [128, 128, 128]);
+
+            $loopCount = 10;
+
+            for ($i = 0; $i < $loopCount; $i++) {
+                static::getDatabase()->createDocument('fullTextSizeTest', new Document([
+                    'string1' => 'string1' . $i,
+                    'string2' => 'string2' . $i,
+                    'string3' => 'string3' . $i,
+                ]));
+            }
+
+            $size2 = static::getDatabase()->getSizeOfCollection('fullTextSizeTest');
+
+            $this->assertGreaterThan($size1, $size2);
+
+            static::getDatabase()->createIndex('fullTextSizeTest', 'fulltext_index', Database::INDEX_FULLTEXT, ['string1']);
+
+            $size3 = static::getDatabase()->getSizeOfCollection('fullTextSizeTest');
+
+            $this->assertGreaterThan($size2, $size3);
+        } else {
+            $this->expectNotToPerformAssertions();
+        }
+    }
```
```suggestion
        // SQLite does not support fulltext indexes
        if (!static::getDatabase()->getAdapter()->getSupportForFulltextIndex()) {
            $this->expectNotToPerformAssertions();
            return;
        }
        
        static::getDatabase()->createCollection('fullTextSizeTest');

        $size1 = static::getDatabase()->getSizeOfCollection('fullTextSizeTest');

        static::getDatabase()->createAttribute('fullTextSizeTest', 'string1', Database::VAR_STRING, 128, true);
        static::getDatabase()->createAttribute('fullTextSizeTest', 'string2', Database::VAR_STRING, 254, true);
        static::getDatabase()->createAttribute('fullTextSizeTest', 'string3', Database::VAR_STRING, 254, true);
        static::getDatabase()->createIndex('fullTextSizeTest', 'index', Database::INDEX_KEY, ['string1', 'string2', 'string3'], [128, 128, 128]);

        $loopCount = 10;

        for ($i = 0; $i < $loopCount; $i++) {
            static::getDatabase()->createDocument('fullTextSizeTest', new Document([
                'string1' => 'string1' . $i,
                'string2' => 'string2' . $i,
                'string3' => 'string3' . $i,
            ]));
        }

        $size2 = static::getDatabase()->getSizeOfCollection('fullTextSizeTest');

        $this->assertGreaterThan($size1, $size2);

        static::getDatabase()->createIndex('fullTextSizeTest', 'fulltext_index', Database::INDEX_FULLTEXT, ['string1']);

        $size3 = static::getDatabase()->getSizeOfCollection('fullTextSizeTest');

        $this->assertGreaterThan($size2, $size3);
    }
```
